### ISSUE TITLE
CC0004 and exception filters

### ISSUE THREAD

**Issue by dariusbuinovskij** [code-cracker/code-cracker#980]
#New analyzer:

CC0004 analyzer unable to handle filters. I some cases I need to log exceptions (catch block never reached). I can disable warning by adding #pragma warning disable, but I would like to avoid that. Can I expect that to be fixed in coming releases?

Before:

````csharp
		static void Main()
		{
			try
			{
				// Some code...
			}
			catch (Exception ex) when (Log(ex))
			{ }
		}

		static bool Log(Exception ex)
		{
			Console.WriteLine(ex.Message);
			return false;
		}

````

After:

````csharp
			try
			{
				// Some code...
			}
#pragma warning disable CC0004 // Catch block cannot be empty
			catch (Exception ex) when (Log(ex))
			{ }
#pragma warning restore CC0004 // Catch block cannot be empty
		}

		static bool Log(Exception ex)
		{
			Console.WriteLine(ex.Message);
			return false;
		}
````

You can add more information here, e.g. conditions under which a diagnostic should not trigger, etc.

Diagnostic Id: `CC0000` (take a number and update the [wiki](https://github.com/code-cracker/code-cracker/wiki/DiagnosticIds))
Category: `<some category>` (see [supported categories](https://github.com/code-cracker/code-cracker/blob/master/src/Common/CodeCracker.Common/SupportedCategories.cs) and their [descriptions](https://github.com/code-cracker/code-cracker/issues/97))
Severity: `Hidden | Info | Warning | Error` (see the [descriptions](https://github.com/code-cracker/code-cracker/#severity-levels))

#Bug

Input your bug description. Make sure you describe the steps to reproduce,
that you are working with the latest version, and the issue has not been reported yet.

Example: (don't use your project code, use a sample that anyone could use to verify the bug,
so, for example,
don't use classes that are not part of the BCL or declared on your sample.)

````csharp
//the code that reproduces the bug
````

Current output after fix applied (if it is a code fix bug):

````csharp
//code fixed incorrectly
````

Expected output after fix applied (if it is a code fix bug):

````csharp
//code fixed incorrectly
````

**Comment by jwooley** [comment]
At first glance I was going to recommend against this as a typical monadic comprehension (like filter/where) would not guard against side effects. However, checking the [C# 6 FAQ,](https://blogs.msdn.microsoft.com/csharpfaq/2014/11/20/new-features-in-c-6/) this behavior is actually recommended:

> It is also a common and accepted form of “abuse” to use exception filters for side effects; e.g. logging. They can inspect an exception “flying by” without intercepting its course. In those cases, the filter will often be a call to a false-returning helper function which executes the side effects:

This may be a valid extension to make for this analyzer. We could go even further and have an analyzer that detects the following pattern and recommends using an expression filter to avoid the catch/throw performance penalty:

bad:

```
try
{
  // Do something
}
catch
{
  WriteLine(ex.Message);
  throw;
}
```

```
better:

```
try
{
  // Do Something
}
catch (Exception ex) when (Log(ex))
{
}

bool Log(Exception ex)
{
  WriteLine(ex.Message);
  // TODO: Add conditional evaluation logic here
  return false;
}

```

**Comment by dariusbuinovskij** [comment]
Yes, I know that logging is side effect of exception filters, but is better than
`try
{
  // Do something
}
catch
{
  WriteLine(ex.Message);
  throw;
}`

Yes, analyzer you suggested sounds good. As I understand error about empty catch block will disappear as well when you will add such extra analyzer?
